\documentclass[10pt]{article}
\usepackage{hyperref}
\usepackage{booktabs}
\usepackage{ltablex}
\usepackage[dvipsnames]{xcolor}
\usepackage{adjustbox}
\usepackage{titlesec}
\usepackage{tabularx}

\usepackage[headsep=0.05in,headheight=4in,margin=0.8in]{geometry}

\usepackage{fancyhdr}

\usepackage{graphicx}

\fancyhf{}% Clear header/footer
\setlength{\headwidth}{\textwidth}
\fancyhead[R]{\color{NavyBlue}\textbf{\sffamily \Large Buffalo Institute for Genomics and Data Analytics}}
\fancyhead[L]{ % right
   \includegraphics[trim={5cm 7cm 5cm 5cm},clip,height=0.60in]{\VAR{image_path}}
}
\renewcommand{\footrulewidth}{1pt}% default is 0pt
\fancyfoot[L]{Page \thepage }
\fancyfoot[R]{701 Ellicott Street, Buffalo, NY 14203 $\bullet$ 716-881-7514}
\pagestyle{fancy}
\titlespacing\section{0pt}{12pt plus 2pt minus 2pt}{0pt plus 2pt minus 2pt}
\titleformat*{\section}{\normalfont\fontsize{12}{14}\color{NavyBlue}\mdseries}
\renewcommand{\familydefault}{\sfdefault}
\newcommand\Tstrut{\rule{0pt}{2.5ex}}       % "top" strut
\newcommand\Bstrut{\rule[-0.9ex]{0pt}{0pt}} % "bottom" strut
\newcommand{\TBstrut}{\Tstrut\Bstrut} % top&bottom struts
\begin{document}
\begin{Form}
\begin{table}[h]
\noindent\begin{tabular}{@{}ll}
\TextField[charsize=8pt,width=3.88cm,color=black,bordercolor=1 1 1]{Sample ID \#:} &
\TextField[charsize=8pt,width=3.20cm,color=black,bordercolor=1 1 1]{Date/Time Collected:} \\
\TextField[charsize=8pt,width=3.05cm,color=black,bordercolor=1 1 1]{Requesting Facility:} &
\TextField[charsize=8pt,width=3.28cm,color=black,bordercolor=1 1 1]{Date/Time Received:} \\
\TextField[charsize=8pt,width=3.82cm,color=black,bordercolor=1 1 1]{Requesting PI:} &
Data Release Date: \\
\end{tabular}
\end{table}
\end{Form}
\begin{center}
\Large\textbf{\textsf{\color{NavyBlue}Report}}
\end{center}
\section*{Result Summary}
\noindent \VAR{result_summary}
%% if relevant_findings
\section*{Variants Relevant to Indication for Testing}
\footnotesize
\noindent\begin{tabularx}{\textwidth}{@{}p{0.2in}@{}p{1.2in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}}
    \multicolumn{2}{@{}l}{\textbf{Variant ID }} &
    \textbf{Allele Frequency*} &
    \textbf{Zygosity} &
    \textbf{Gene} &
    \textbf{RefSeq ID} &
    \textbf{Variant} \\
    \cmidrule[1pt](r{1pt}){1-2}
    \cmidrule[1pt](r{1pt}){3-3}
    \cmidrule[1pt](r{1pt}){4-4}
    \cmidrule[1pt](r{1pt}){5-5}
    \cmidrule[1pt](r{1pt}){6-6}
    \cmidrule[1pt](r{1pt}){7-7}
    \endhead
    %% for (id, variant_id, allele_frequency, zygosity, genes, refseq_ids, variants, clinvar, gwas) in relevant_findings
    \multicolumn{2}{@{}l}{\VAR{variant_id}} &
    \VAR{allele_frequency} &
    \VAR{zygosity} &
    %% for gene in genes
        \VAR{gene} \newline
    %% endfor
    &
    %% for refseq_id in refseq_ids
        \VAR{refseq_id} \newline
    %% endfor
    &
    %% for variant in variants
        \VAR{variant} \newline
    %% endfor
    \\
    %% if not refseq_ids and not variants
        \\
    %% endif
    %% if clinvar
        %% for cvar, sig in clinvar
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{ClinVar:}} &
        \multicolumn{5}{@{}m{12cm}}{\VAR{cvar}} \\
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{Classification:}} &
        \multicolumn{5}{@{}l}{\VAR{sig}} \\
        %% endfor
    %% endif
    %% if gwas
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{GWAS Catalog:}} &
        \multicolumn{5}{@{}m{12cm}}{\VAR{gwas}} \\
    %% endif
    \cmidrule{1-7}
    %% endfor
\end{tabularx}
%% endif
%% if incidental_findings
\section*{Other Variants of Medical Significance (Incidental Findings)}
\footnotesize
\noindent\begin{tabularx}{\textwidth}{@{}p{0.2in}@{}p{1.2in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}@{}p{1.06in}}
    \multicolumn{2}{@{}l}{\textbf{Variant ID}} &
    \textbf{Allele Frequency*} &
    \textbf{Zygosity} &
    \textbf{Gene} &
    \textbf{RefSeq ID} &
    \textbf{Variant} \\
    \cmidrule[1pt](r{1pt}){1-2}
    \cmidrule[1pt](r{1pt}){3-3}
    \cmidrule[1pt](r{1pt}){4-4}
    \cmidrule[1pt](r{1pt}){5-5}
    \cmidrule[1pt](r{1pt}){6-6}
    \cmidrule[1pt](r{1pt}){7-7}
    \endhead
    %% for (id, variant_id, allele_frequency, zygosity, genes, refseq_ids, variants, clinvar, gwas) in incidental_findings
    \multicolumn{2}{@{}l}{\VAR{variant_id}} &
    \VAR{allele_frequency} &
    \VAR{zygosity} &
    %% for gene in genes
        \VAR{gene} \newline
    %% endfor
    &
    %% for refseq_id in refseq_ids
        \VAR{refseq_id} \newline
    %% endfor
    &
    %% for variant in variants
        \VAR{variant} \newline
    %% endfor
    \\
    %% if not refseq_ids and not variants
        \\
    %% endif
    %% if clinvar
        %% for cvar, sig in clinvar
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{ClinVar:}} &
        \multicolumn{5}{@{}m{12cm}}{\VAR{cvar}} \\
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{Classification:}} &
        \multicolumn{5}{@{}l}{\VAR{sig}} \\
        %% endfor
    %% endif
    %% if gwas
        \hfill &
        \multicolumn{1}{@{}l}{\textbf{GWAS Catalog:}} &
        \multicolumn{5}{@{}m{12cm}}{\VAR{gwas}} \\
    %% endif
    \cmidrule{1-7}
    %% endfor
\end{tabularx}
%% endif
\normalsize

\section*{Methodology}
\noindent \VAR{methodology}

\bigskip
\noindent
\footnotesize
\def\arraystretch{1.5}
\noindent\begin{tabular}{@{}|p{1.65in}|l|l|l|l|}
  \hline
  \textbf{Variant} & \textbf{SIFT Prediction} & \textbf{LRT Prediction} & \textbf{Polyphen2 HDIV Prediction} & \textbf{Polyphen2 HVAR Prediction} \\ \hline
  %% for (variant, sift, lrt, phdiv, phvar) in predictor_array
  \VAR{variant} & \VAR{sift} & \VAR{lrt} & \VAR{phdiv} & \VAR{phvar} \\ \hline
  %% endfor
\end{tabular}

\bigskip
\noindent The following table shows the interpretation of the predictions:
\newline
\bigskip
\newline
\footnotesize
\def\arraystretch{1.5}
\begin{tabular}{|l|p{5.75in}|}
  \hline
  \textbf{Predictor Name} & \textbf{Description} \\ \hline
  SIFT & D: Deleterious (sift score $\leq 0.05$); T: Tolerated (sift score $> 0.05$) \\ \hline
  LRT & D: Deleterious; N: Neutral; U: Unknown \\ \hline
  Polyphen2 HDIV & D: probably damaging (hdiv $\geq 0.957$); P: possibly damaging ($0.453 \leq$ hdiv $\leq 0.957$); B: benign (hdiv $\leq 0.957$) \\ \hline
  Polyphen2 HVAR & D: probably damaging (hvar $\geq 0.909$); P: possibly damaging ($0.447 \leq$ hvar $\leq 0.909$); B: benign (hvar $\leq 0.446$) \\ \hline
\end{tabular}
\bigskip
\newline
\noindent*Allele frequency is based on Exome Aggregation Consortium (ExAC) all available populations
\normalsize
\bigskip

\section*{Additional Notes}
\noindent \VAR{additional_notes}
\bigskip
\newline \noindent Report generated: \today

\end{document}
