{% extends "base.html" %}
{% load staticfiles %}

{% block title %}
       Tutorials
{% endblock %}

{% block content %}
<script>
    $("#main-nav-bar > ul > li.active").removeClass("active")
    $("#navbarId-Help").addClass("active")
    $("#navbarId-Tutorial").addClass("active")
</script>

<div id="top" class="page-header">
    <h1>Tutorials</h1>
</div>

<ul>
    <li><a href="#usecase1">Use Case 1</a></li>
    <li><a href="#usecase2">Use Case 2</a></li>
    <li><a href="#usecase3">Use Case 3</a></li>
</ul>

<hr>
<div id="usecase1" class="container paragraph-justify bold-hr">

  <div class="panel-group">
    <div class="panel panel-info">
      <div class="panel-heading">
        <h3 class="panel-title">
          <a data-toggle="collapse" href="#tutorial1">Use Case 1</a>
        </h3>
      </div>
      <div id="tutorial1" class="panel-collapse collapse in">
        <div class="panel-body">
            <div class="well"><p><strong>Use Case 1:</strong>
                Retrieve variants from the Brain tumor (GBM from TCGA) study that are predicted to be pathogenic by the SIFT and Polyphen2 algorithms, with MAF &lt;= 0.001 in all of the populations from ExAC. Return variant names, genomic coordinates, genes affected, samples that harbor these variants and genotypes for these variants. Save results to a tab-delimited text file.
            </p>
            </div>
            <hr>
              <div class="panel-body">
                    <div class="well">To begin searching, click the green <strong>"Start Searching"</strong> button. You will be forwarded to a page that asks you to select study and dataset with two drop-down menus.</div>
                    <div class="col-md-12">
                        <img src="{% static 'custom/images/tutorial/tut1_1.png'%}" class="img-responsive">
                    </div>
              </div>
              <hr>

              <div class="panel-body">
                    <div class="well">
                        Select <strong>"Brain tumor (GBM, from TCGA, demo only)"</strong> and <strong>"Somatic Mutations, GRCh37/hg19"</strong> for study and dataset, respectively. Click the <strong>"Next"</strong> button to proceed to the filter selection page.
                    </div>

                        <img src="{% static 'custom/images/tutorial/tut1_2.png'%}" class="img-responsive">
              </div>
              <hr>

              <div class="panel-body">

                    <div class="well">
                        You will be presented with the <strong>"Select Filters"</strong> page.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_3.png'%}" class="img-responsive">


              </div>
              <hr>

              <div class="panel-body">
                    <div class="well">
                        Click the <strong>"Variant related information"</strong> text in the blue panel heading to collapse this section because we will not be choosing any filters from it.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_4.png'%}" class="img-responsive">
              </div>
              <hr>

              <div class="panel-body">
                    <div class="well">
                        Click the  <strong>"Pathogenic prediction"</strong> text in the blue panel heading to expand this section. Select <strong>"Deleterious"</strong> for <strong>"SIFT pred"</strong>. Next select both <strong>"probably damaging"</strong> and <strong>"possibly damaging"</strong> for <strong>"Polyphen2 HDIV pred"</strong> by ctrl-clicking in the multi-select box. Hover over the information icon to the right of a filter name if you'd like more information regarding a filter.
                    </div>
                    <img src="{% static 'custom/images/tutorial/tut1_5.png'%}" class="img-responsive">

              </div>
              <hr>

              <div class="panel-body">
                    <div class="well">
                        Click the  <strong>"Minor allele frequency"</strong> text in the blue panel heading to expand this section. Enter value <strong>"0.001"</strong> for all the <strong>"ExAC"</strong> filters except <strong>"ExAC All"</strong>. The selected filters also show up in the right panel, where you can more quickly remove them if you desire. Click the <strong>"Next"</strong> button to proceed to the attribute selection page.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_6.png'%}" class="img-responsive">
              </div>
                <hr>

              <div class="panel-body">
                    <div class="well">
                        In the attribute selection page, you can select the attributes that you want to include in the search result. Check the attribute names <strong>"Variant ID"</strong>, <strong>"Chromosome"</strong>, <strong>"Start Position"</strong>, <strong>"Reference allele"</strong>, and <strong>"Alternate allele"</strong> under the <strong>"Variant Related Information"</strong> section. Then click the blue panel heading to collapse it.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_7.png'%}" class="img-responsive">
              </div>
                <hr>

              <div class="panel-body">
                    <div class="well">
                        Expand the <strong>"Gene/Transcript"</strong> section and select <strong>"Gene"</strong> and <strong>"RefGene ID"</strong>under the <strong>"RefSeq Identifier"</strong> sub-section. Click the blue panel heading to collapse it.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_8.png'%}" class="img-responsive">
              </div>
                <hr>

              <div class="panel-body">
                    <div class="well">
                        Click the <strong>"Patient/Sample Related Information"</strong> panel title text to expand it. Click the <strong>"Select All"</strong> button to choose all 3 attributes. Click the green <strong>"Search"</strong> button to begin searching.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_9.png'%}" class="img-responsive">
              </div>
            <hr>

              <div class="panel-body">
                    <div class="well">
                        Results are sortable by column. By default, up to 400 search results are shown as a preview. If you wish to modify your search/ output criteria, you can use the circle tabs that are above the search results to go back to the filter/ attribute pages. To download your results, you can click the <strong>"Download"</strong> button at the top-right corner to save the search result as a tab-delimited text file.
                    </div>
                        <img src="{% static 'custom/images/tutorial/tut1_10.png'%}" class="img-responsive">

              </div>


        </div>
      </div>
    </div>
  </div>
</div>
<a href="#top">Go to top</a>
<hr>

<div id="usecase2" class="container paragraph-justify bold-hr">
  <div class="panel-group">
    <div class="panel panel-info">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" href="#tutorial2">Use Case 2</a>
        </h4>
      </div>
      <div id="tutorial2" class="panel-collapse collapse in">
        <div class="panel-body">
            <div class="well">
            <p><strong>Use Case 2:</strong>
                A researcher is interested in studying the <strong>"NFIX"</strong> gene,
                which has been shown to affect muscle development. They are interested in identifying non-silent variants in <strong>"NFIX"</strong> as well as other genes from the nuclear factor 1 family, such as <strong>"NFIA"</strong>, <strong>"NFIB"</strong> and <strong>"NFIC"</strong> genes from the samples in the Brain tumor (GBM from TCGA) dataset.
            </p>
            </div>
            <hr>
            <div class="well">Begin searching, by selecting the <strong>"Brain tumor (GBM, from TCGA, demo only)"</strong> study and the <strong>"Somatic Mutations, GRCh37/hg19"</strong> dataset.
            </div>
                <img src="{% static 'custom/images/tutorial/tut1_2.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                In the filter selection page, expand the <strong>Genes/Transcript</strong> section and under the <strong>"RefSeq identifier"</strong> sub-section, specify Genes: <strong>"NFIX"</strong>, <strong>"NFIA"</strong>, <strong>"NFIB"</strong>, and <strong>"NFIC"</strong>.
            </div>
                <img src="{% static 'custom/images/tutorial/tut2_2.png'%}" class="img-responsive">

              <hr>


            <div class="well">
                Expand the <strong>Functional Consequence</strong> section and under the <strong>"Consequence to RefSeq gene/transcript"</strong> sub-section, select the following choices: <strong>"nonsynonymous SNV"</strong>, <strong>"splicing"</strong>, <strong>"stopgain"</strong> and <strong>"stoploss"</strong>. You can select multiple choices by holding down the ctrl key and making the selections.
            </div>
                <img src="{% static 'custom/images/tutorial/tut2_3.png'%}" class="img-responsive">
              <hr>

               <div class="well">
                Expand the <strong>"Minor allele Frequency (MAF)"</strong> section and under the <strong>"The Exome Aggregation Consortium (ExAC)"</strong> sub-section, input "0.05" for <strong>"ExAC ALL (&lt;)"</strong>. Click the <strong>"Next"</strong> button to proceed to the attribute selection page.
            </div>
                <img src="{% static 'custom/images/tutorial/tut2_4.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                From the <strong>"Variant Related Information"</strong> section select the following attributes: <strong>"Chromosome"</strong>, <strong>"Start Position"</strong>, <strong>"Reference allele"</strong>, and <strong>"Alternate allele"</strong>. Collapse the current section by clicking the blue panel heading.

            </div>
                <img src="{% static 'custom/images/tutorial/tut2_5.png'%}" class="img-responsive">
              <hr>

             <div class="well">
                Expand the <strong>"Functional Consequence"</strong> section. Select <strong>"Consequence to Ensembl gene/transcript"</strong> attribute. Collapse the current section by clicking the blue panel heading.

            </div>
                <img src="{% static 'custom/images/tutorial/tut2_6.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                Expand the <strong>"Gene/Transcript"</strong> section. From the "RefSeq Identifier" sub-section, select <strong>"Gene"</strong> attribute. Collapse the current section by clicking its title text in blue panel heading.

            </div>
                <img src="{% static 'custom/images/tutorial/tut2_7.png'%}" class="img-responsive">
              <hr>


            <div class="well">
                Expand the <strong>"Patient/Sample Related Information"</strong> section. Select the attributes <strong>"Sample genotype"</strong> and <strong>"Sample ID"</strong>. Click the green <strong>"Search"</strong> button to begin searching.

            </div>
                <img src="{% static 'custom/images/tutorial/tut2_8.png'%}" class="img-responsive">
              <hr>


            <div class="well">
                Results show that the demo dataset has 5 matches.

            </div>
                <img src="{% static 'custom/images/tutorial/tut2_9.png'%}" class="img-responsive">
              <hr>

        </div>
      </div>
    </div>
  </div>
</div>
<a href="#top">Go to top</a>
<hr>

<div id="usecase3" class="container paragraph-justify bold-hr">
  <div class="panel-group">
    <div class="panel panel-info">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" href="#tutorial3">Use Case 3</a>
        </h4>
      </div>
      <div id="tutorial3" class="panel-collapse collapse in">
        <div class="panel-body">
            <div class="well">
            <p><strong>Use Case 3:</strong>
                Review variants in a gene of interest to ascertain whether they are true calls or false positive calls. For each variant reviewed, assign an appropriate status. Use a post-query filter to view only variants which have been judged to be true calls.
            </p>
            </div>
            <hr>
            <div class="well">Begin searching, by selecting the <strong>"Statin Induced Myopathy"</strong> study and the <strong>"SIM Case (WGS) and Control (WES), GRCh37/hg19"</strong> dataset. Click <strong>"Next"</strong> to proceed to filter selection.
            </div>
                <img src="{% static 'custom/images/tutorial/tut3_1.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                In the filter selection page, expand the <strong>Genes/Transcript</strong> section and under the <strong>"RefSeq identifier"</strong> sub-section, specify Gene <strong>"KLHL17"</strong>. Expand the <strong>Functional Consequence</strong> section and under the <strong>"Consequence to RefSeq gene/transcript"</strong> sub-section, select <strong>"nonsynonymous SNV"</strong>. Click the <strong>"Next"</strong> button to proceed to the attribute selection page.
            </div>
                <img src="{% static 'custom/images/tutorial/tut3_2.png'%}" class="img-responsive">

              <hr>

            <div class="well">
                From the <strong>"Variant Related Information"</strong> section select the following attributes: <strong>"Variant ID"</strong>, <strong>"GATK Filter"</strong>, and <strong>"GATK VQSR Score"</strong>. Click the green <strong>"Search"</strong> button to begin searching.
            </div>
                <img src="{% static 'custom/images/tutorial/tut3_3.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                There should be 6 results available. Click the <strong>"Show Review Status"</strong> button to bring up a <strong>"Review Status"</strong>  column on the left side of the results table. This allows you to approve, reject, or mark as questionable any of the results.

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_4.png'%}" class="img-responsive">
              <hr>

             <div class="well">
                Click the Variant ID <strong>"1-896105-G-T"</strong> and a new browser tab will open to that variant's Variant Page. This page contains all the information available on that particular call, along with links to various databases and viewing tools. Following the links and analyzing the data they provide can lead to a decision on whether or not this is a true variant.

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_5.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                Return to the Search tab in your browser. For demonstration purposes, let's assume 2 of the results are approved (green checkmark), 2 are rejected (red 'x'), and 2 are pending (orange question mark), as in the image below.

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_6.png'%}" class="img-responsive">
              <hr>


            <div class="well">
                You can now use the <strong>"Select Post Query Filter"</strong> button to further filter the results. Click the button and select <strong>"Rejected"</strong> and <strong>"Pending"</strong>, then click <strong>"Apply"</strong>.

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_7.png'%}" class="img-responsive">
              <hr>


            <div class="well">
                The results table now shows only those variants which not only adhere to the initial filtering but also the additional post-query filtering. The variants which only pass the initial filtering are listed in a separate table below the main one.

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_8.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                The variants which have been assigned a review status can be managed by clicking <strong>"Variant Review Status Summary"</strong> under the <strong>Search</strong> menu. This page is unique per user group. 

            </div>
                <img src="{% static 'custom/images/tutorial/tut3_9.png'%}" class="img-responsive">
              <hr>

            <div class="well">
                Each category of variant review status can be inspected and individual entries can have their review status changed by clicking the category's <strong>"Edit/show more"</strong> link. Along with the ability to edit review status, you can also view the history of review status changes on these pages.
            </div>
                <img src="{% static 'custom/images/tutorial/tut3_10.png'%}" class="img-responsive">
              <hr>

        </div>
      </div>
    </div>
  </div>
</div>
<a href="#top">Go to top</a>

{% endblock %}
